src/edu/csus/ecs/pc2/clics/API202306/EventFeedService.java

```diff
@@ -0,0 +1,183 @@
+        final ServletOutputStream servletOutputStream = asyncContext.getResponse().getOutputStream();
+
+        if (eventFeedStreamer == null) {
+            eventFeedStreamer = new EventFeedStreamer(contest, controller, servletRequest, sc);
+        }
+
+        EventFeedFilter filter = new EventFeedFilter();
+        filter.setClient(servletRequest.getRemoteUser() + "@" + servletRequest.getRemoteAddr() + ":" + servletRequest.getRemotePort());
+
+        String startMsg = "Starting Event Feeder for " + filter.getClient() + ": ";
+
+        if (eventTypeList != null) {
+            filter.addEventTypeList(eventTypeList);
+            startMsg += "sending only event types '" + eventTypeList + "' ";
+        }
+
+        if (startingToken != null) {
+            if (startingToken.startsWith(EventFeedJSON.EVENT_ID_PREFIX) && Utilities.isIntegerNumber(startingToken.substring(EventFeedJSON.EVENT_ID_PREFIX_LENGTH))) {
+                filter.addStartingEventId(startingToken);
+                startMsg += "starting after id " + startingToken;
+            } else {
+                response.sendError(HttpServletResponse.SC_BAD_REQUEST, "Not starting event feed: Invalid starting id: `"+startingToken+"`");
+                return;
+            }
+        } else {
+            startMsg += "(whole feed)";
+        }
+        info(startMsg);
+
+        /**
+         * Add stream and write past events to stream.
+         */
+        eventFeedStreamer.addStream(servletOutputStream, filter);
+
+        if (!eventFeedStreamer.isRunning()) {
+            /**
+             * Put on thread if not running on a thread.
+             */
+            new Thread(eventFeedStreamer).start();
+        }
+
+        while (true) {
+            // If the contest has been finalized or this client is not longer connected (eg not receiving the EF), we are done
+            if (eventFeedStreamer.isFinalized() || !eventFeedStreamer.isStreamActive(servletOutputStream)) {
+                break;
+            }
+            try {
+                Thread.sleep(1 * Constants.MS_PER_SECOND);
+            } catch (InterruptedException e) {
+                e.printStackTrace();
+                log.log(Level.WARNING, "During sleep " + e.getMessage());
+                break;
+            }
+        }
+        info("Event Feeder client " + filter.getClient() + " is no longer connected.");
+    }
+
+    public void info(String message) {
+        if (log != null) {
+            log.log(Level.INFO, message);
+        }
+    }
+
+    /**
+     * Retrieve access information about this endpoint for the supplied user's security context
+     *
+     * @param sc User's security information
+     * @return CLICSEndpoint object if the user can access this endpoint's properties, null otherwise
+     */
+    public static CLICSEndpoint getEndpointProperties(SecurityContext sc) {
+        String [] efProps = { "type", "id", "data", "token" };
```

Comment:
It seems strange that these `efProps` are defined here as hard-coded strings; I would expect them to be defined as constants (somewhere).  Also, I can't find any definition of these properties anywhere in the CLICS spec (some -- but not all -- of them appear in the `Examples` section under the `Event feed` endpoint description, but I can't find anywhere that they are actually defined/listed in the spec).  (This of course is a problem with the spec, not with this code... but still I'm trying to figure out the source of their definition...)

Reply:
I think I answered this in the slack chat, but I will duplicate that response here (without my editorial) for posterity:

> The 2023-06 spec has it buried like other things in the spec.  See "General Design Principles"->Notification Format. https://ccs-specs.icpc.io/2023-06/contest_api?highlight=num_solved#notification-format 

I'm OK with having these as string literals in the array since they are specific to the protocol being implemented, and, are local to this module.